packages/-ember-data/tests/integration/record-data/record-data-test.ts

```diff
@@ -0,0 +1,310 @@
+    let calledPush = 0
+    let calledClientDidCreate = 0;
+    let calledWillCommit = 0;
+    let calledWasRejected = 0;
+    let calledUnloadRecord = 0;
+    let calledRollbackAttributes = 0;
+    let calledDidCommit = 0;
+
+    class LifecycleRecordData extends TestRecordData {
+      pushData(data, calculateChange?: boolean) {
+        calledPush++;
+      }
+
+      clientDidCreate() {
+        calledClientDidCreate++;
+      }
+
+      willCommit() {
+        calledWillCommit++;
+      }
+
+      commitWasRejected() {
+        calledWasRejected++;
+      }
+
+      unloadRecord() {
+        calledUnloadRecord++;
+      }
+
+      rollbackAttributes() {
+        calledRollbackAttributes++;
+      }
+
+      didCommit(data) {
+        calledDidCommit++;
+      }
+    }
+
+
+    let TestStore = Store.extend({
+      createRecordDataFor(modelName, id, clientId, storeWrapper) {
+        return new LifecycleRecordData();
+      }
+    });
+
+    let TestAdapter = Ember.Object.extend({
+      updateRecord() {
+        called++;
+        if (called === 1) {
+          return Promise.resolve();
+        } else if (called > 1) {
+          return Promise.reject();
+        }
+      },
+
+      createRecord() {
+        return Promise.resolve();
+      }
+    });
+
+    owner.register('service:store', TestStore);
+    owner.register('adapter:application', TestAdapter, { singleton: false });
+
+    store = owner.lookup('service:store');
+
+    store.push({
+      data: [personHash]
+    });
+
+    let person = store.peekRecord('person', '1');
+    person.save();
+    await settled();
+    person.save();
+    await settled();
+    person.rollbackAttributes();
+    person.unloadRecord();
+    await settled();
+    assert.equal(calledPush, 1, 'Called pushData');
+    assert.equal(calledWillCommit, 2, 'Called willCommit');
+    assert.equal(calledWasRejected, 1, 'Called commitWasRejected');
+    assert.equal(calledUnloadRecord, 1, 'Called unloadRecord');
+    assert.equal(calledRollbackAttributes, 1, 'Called rollbackAttributes');
+    assert.equal(calledDidCommit, 1, 'Called didCommit');
+    assert.equal(calledClientDidCreate, 0, 'Did not called clientDidCreate');
+
+    calledPush = 0;
+    calledClientDidCreate = 0;
+    calledWillCommit = 0;
+    calledWasRejected = 0;
+    calledUnloadRecord = 0;
+    calledRollbackAttributes = 0;
+    calledDidCommit = 0;
+
+    let clientPerson = store.createRecord('person', { id: 2 });
+    clientPerson.save();
+    await settled();
+    clientPerson.save();
+    await settled();
+    clientPerson.unloadRecord();
+    await settled();
+    assert.equal(calledPush, 0, 'Called pushData');
+    assert.equal(calledWillCommit, 2, 'Called willCommit');
+    assert.equal(calledWasRejected, 1, 'Called commitWasRejected');
+    assert.equal(calledUnloadRecord, 1, 'Called unloadRecord');
+    assert.equal(calledDidCommit, 1, 'Called didCommit');
+    assert.equal(calledClientDidCreate, 1, 'Did not called clientDidCreate');
```

Comment:
I would add something to check the ordering of these calls over just asserting counts, because as written `clientDidCreate` could be called after the first successful save (or later) and we'd never know if that was the right intent.

Typically asserting just after the code that invokes the counter is a safer way to ensure the timing.

Reply:
Great point